vowpalwabbit/bfgs.cc

```diff
@@ -457,18 +457,19 @@ double add_regularization(vw& all, bfgs& b, float regularization, T& weights)
     }
 
   // if we're not regularizing the intercept term, then subtract it off from the result above
+  // when accessing weights[constant], always use weights.strided_index(constant)
   if (all.no_bias)
   {
     if (b.regularizers == nullptr)
     {
-      (&weights[constant])[W_GT] -= regularization * weights[constant];
-      ret -= 0.5 * regularization * (weights[constant]) * (weights[constant]);
+      (&weights.strided_index(constant))[W_GT] -= regularization * (weights.strided_index(constant));
```

Comment:
Is there a use case to NOT using strided_index? If not then perhaps we should make the `operator[]` private to avoid this bug in future?

Reply:
I don't know vw well enough to comment on that but like we discussed in our conversation earlier today, it would be a big change, probably out of scope of this PR

Reply:
This is definitely out of scope of the PR and if we decide we want to look into this I can create and issue to track it

Reply:
Unfortunately, there is.  Currently, in setup_example all features are multipled by a stride which allows direct access to the parameters via [] as per https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/gd_predict.h#L33

Reply:
So does offset correspond with the stride? Maybe the default should be to access using the stride and then explicitly state when you want unstrided? It depends on what the more common/intuitive usage is

Reply:
Not exactly.  

Stride controls floats per parameter.  Offset indexes into logically distinct sets of parameters.